File: CodeTraining/LeetCode/easy/linklist/code_203.scala
package LeetCode.easy.linklist

object code_203 {

  class ListNode(var _x: Int = 0) {
    var next: ListNode = null
    var x: Int = _x
  }

  def removeElements(head: ListNode, `val`: Int): ListNode = {
    var pre = new ListNode
    var root = pre
    // 在scala中不能使用head = head.next，因此重新定义结点来代替
    pre.next = head
    var cur = head
    while (cur != null) {
      if (cur.x == `val`) {
        cur = cur.next
        pre.next = cur
      }
      else {
        pre = cur
        cur = cur.next
      }
    }
    root.next
  }
}

File: CodeTraining/LeetCode/easy/linklist/Code_203.java
package LeetCode.easy.linklist;



public class Code_203 {

    private class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public ListNode removeElements(ListNode head, int val) {

//        ListNode dummyHead = new ListNode(-1);
//        dummyHead.next = head;
//        ListNode prev = dummyHead;
//        while(prev.next!=null){
//            if(prev.next.val == val){
//                prev.next = prev.next.next;
//            }else{
//                prev = prev.next;
//            }
//        }
//        return dummyHead;

        while(head != null && head.val == val){
            ListNode delNode = head;
            head = head.next;
            delNode.next = null;
        }

        if(head == null){
            return null;
        }

        ListNode prev = head;
        while(prev.next != null){
            if(prev.next.val == val){
                prev.next = prev.next.next;
            }else{
                prev = prev.next;
            }
        }
        return head;
    }

}
